Improve detection of the type from a PHP variable (#2814)
doctrine/mongodb-odm: docs/en/reference/custom-mapping-types.rst

        <field field-name="field" type="date_with_timezone" />

Custom Type Example: Mapping a UUID Class
========================================

You can create a custom mapping type for your own value objects or classes. For

> example, to map a UUID value object using the `ramsey/uuid library`_, you can

Reviewer: Nice to see the original use case from the issue as an example here

Author: I was tempted to modify the example to use `symfony/uid` instead. I've kept this example so as not to be too partisan.

Reviewer: I'm a ramsey/uuid traditionalist anyway 😎

> implement a type that converts between your class and the MongoDB binary format.

Reviewer 2: Should this say "BSON binary format", or do you think that'd just be more confusing for users? OK to leave as-is.

> This approach works for any custom class—just adapt the conversion logic to your needs.

Reviewer: ```suggestion
This approach works for any custom class — just adapt the conversion logic to your needs.
```

Reviewer 2: See also: https://www.merriam-webster.com/grammar/em-dash-en-dash-how-to-use

Typically wouldn't have spaces on either side of the em dash. Alternatively, you can rephrase this as:

> This approach works for any custom class by adapting the conversion logic to your needs.

"Just" may be considered subjective language, the same way we wouldn't want to say "easily" or "simply".

Reviewer: Oh my bad, I always write it like that. Maybe it's the GH formatting of this em dash as well. It read to me like a hyphen cause it looks so short

Author: In French we would add space around the dash in that context too. I recognize that it was copilot who wrote this text.

Example Implementation (using Ramsey\Uuid\UuidInterface)::

.. code-block:: php
    <?php

    namespace My\Project\Types;

    use Doctrine\ODM\MongoDB\Types\ClosureToPHP;
    use Doctrine\ODM\MongoDB\Types\Type;
    use InvalidArgumentException;
    use MongoDB\BSON\Binary;
    use Ramsey\Uuid\Uuid;
    use Ramsey\Uuid\UuidInterface;

    final class UuidType extends Type
    {
        use ClosureToPHP;

        public function convertToPHPValue(mixed $value): ?Uuid
        {
            if (null === $value) {
                return null;
            }

            if ($value instanceof Uuid) {
                return $value;
            }

            if ($value instanceof Binary) {
                return Uuid::fromBytes($value->getData());
            }

            if (is_string($value) && Uuid::isValid($value)) {
                return Uuid::fromString($value);
            }

            throw new InvalidArgumentException(
                sprintf(
                    'Could not convert database value "%s" from "%s" to %s',
                    $value,

>                     get_debug_type($value),

Reviewer 2: Are `use function` lines necessary? I'm OK to omit them, but just want to ensure you've considered the question.

Author: The `use function` are not necessary for global functions. That's a CS preference to use them. I omit them for brevity in the example.

                    UuidInterface::class
                )
            );
        }

        public function convertToDatabaseValue(mixed $value): ?Binary
        {
            if (null === $value || [] === $value) {
                return null;
            }

            if ($value instanceof Binary) {
                return $value;
            }

            if (is_string($value) && Uuid::isValid($value)) {
                $value = Uuid::fromString($value)->getBytes();
            }

            if ($value instanceof Uuid) {
                return new Binary($value->getBytes(), Binary::TYPE_UUID);
            }

            throw new InvalidArgumentException(
                sprintf(
                    'Could not convert database value "%s" from "%s" to %s',
                    $value,
                    get_debug_type($value),
                    Binary::class
                )
            );
        }
    }

Usage Example::

.. code-block:: php

>     #[Field(type: Uuid::class)]

Author: Setting the custom type here should not be necessary. As soon as you register a type with the value object type name, it should be registered automatically https://github.com/doctrine/mongodb-odm/blob/afb565e6246d2dcdaa81636e5a0020d21c1d6fe5/lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php#L2771-L2793

Reviewer 2: Is it clear from this example that you're referencing `Ramsey\Uuid\Uuid`? There's no `use` statement and this is a separate code block from the custom type class above.

Author: Updated to use the FQCN.

    public Uuid $id;

Register the type in your bootstrap code::

.. code-block:: php

    Type::addType(Ramsey\Uuid\Uuid::class, My\Project\Types\UuidType::class);

.. note::

>    The type name should match the support class name (e.g., ``uuid`` for ``UuidType``) to enable automatic association with the field type.

Reviewer: For the automatic association, how does that work? Does it take everything before the `Type` suffix? How does it handle casing, e.g. in `FooBarType`?

Reviewer 2: I'm not certain, but I think this might utilize [doctrine/inflector](https://github.com/doctrine/inflector) internally.

Reviewer: I think it would be helpful to add here what format would constitute a match.

Author: In fact, there is no such magic. The name should be the class name of the value object.

> This pattern can be adapted for any custom class—just implement the conversion logic for your value object.

Reviewer 2: Same feedback here as @paulinevos' thread above re: "class-just"

Author: This sentence was duplicate if the previous one. I removed it.

> .. |FQCN| raw:: html <abbr title="Fully-Qualified Class Name">FQCN</abbr>

Reviewer 2: @GromNaN: this looks like it may be a false positive, since this is some very niche RST syntax

> .. |FQCN| raw:: html <abbr title="Fully-Qualified Class Name">FQCN</abbr>

Author: I revert changes on this line. I'm not sure how this will be formatted.

        <field field-name="field" type="date_with_timezone" />

> Custom Type Example: Mapping a UUID Class

Reviewer 3: Note: I'm currently working on adding a type for Symfony UUIDs as I needed them for a project. That would make this entire chapter obsolete.

Author: The chapter is about mapping a DTO with a custom type. UUID is only an example, we can find an other example.

> =========================================

Author: ```suggestion
-----------------------------------------
```

You can create a custom mapping type for your own value objects or classes. For
example, to map a UUID value object using the `ramsey/uuid library`_, you can
implement a type that converts between your class and the BSON Binary UUID format.

This approach works for any custom class by adapting the conversion logic to your needs.

Example Implementation (using ``Ramsey\Uuid\Uuid``)::

.. code-block:: php

    <?php

    namespace My\Project\Types;

    use Doctrine\ODM\MongoDB\Types\ClosureToPHP;
    use Doctrine\ODM\MongoDB\Types\Type;
    use InvalidArgumentException;
    use MongoDB\BSON\Binary;
    use Ramsey\Uuid\Uuid;
    use Ramsey\Uuid\UuidInterface;

    final class UuidType extends Type
    {
        // This trait provides default closureToPHP used during data hydration
        use ClosureToPHP;

        public function convertToPHPValue(mixed $value): ?Uuid
        {
            if (null === $value) {
                return null;
            }

            if ($value instanceof Uuid) {
                return $value;
            }

>             if ($value instanceof Binary) {

Reviewer 2: Would it make sense to restrict this to a [MongoDB\BSON\Binary](https://www.php.net/manual/en/class.mongodb-bson-binary.php) where `getType()` returns either `TYPE_OLD_UUID` or `TYPE_UUID`? As-is, it's not clear what would happen if you encountered binary data from the server with a different sub-type.

I'll note that `convertToDatabaseValue()` yields a Binary object with `TYPE_UUID`, so perhaps there's no reason to accept `TYPE_OLD_UUID` here.

If you are intentionally being liberal in what you accept and conservative in what you emit ([Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle)) and would like to leave this as-is, that's fine by me. You can optionally add a comment here explaining that design decision for the benefit of readers.

                return Uuid::fromBytes($value->getData());
            }

            if (is_string($value) && Uuid::isValid($value)) {
                return Uuid::fromString($value);
            }

            throw new InvalidArgumentException(
                sprintf(
                    'Could not convert database value "%s" from "%s" to %s',
                    $value,
                    get_debug_type($value),
                    UuidInterface::class
                )
            );
        }

        public function convertToDatabaseValue(mixed $value): ?Binary
        {
            if (null === $value || [] === $value) {
                return null;
            }

>             if ($value instanceof Binary) {

Reviewer 2: Would you want to ensure you always yield a Binary with sub-type `TYPE_UUID`? If so, you might need to split this into two separate conditionals.

```php
if ($value instanceof Binary && $value->getType() !== Binary::TYPE_UUID) {
    // Note: this may throw if the data is not exactly 16 bytes
    return new Binary($value->getBytes(), Binary::TYPE_UUID);
}

if ($value instanceof Binary) {
    return $value;
}
```

Author: I'd rather simplify the example.
https://github.com/doctrine/mongodb-odm/pull/2814/files#r2393846289

                return $value;

>             }

Author: I can simplify the example and remove this edge cases.
```suggestion
            if (null === $value) {
                return null;
            }
```

Reviewer 2: Without the `$value instanceof Binary` check, wouldn't the method then throw InvalidArgumentException for a Binary value? Or are you comfortable assuming that a Binary would never be passed in?

I didn't follow why `[]` was originally being handled alongside `null`, so I don't have much thought on that change.

Author: In the example of document class, the property has the `?\Ramsey\Uuid\Uuid` type, which ensure the `RamseyUuidType::convertToDatabaseValue()` method never get anything else.
As soon as the field value is the database is not corrupted, the simple implementation will work perfectly. No need to add support for edge cases that are never encountered.

            if (is_string($value) && Uuid::isValid($value)) {
                $value = Uuid::fromString($value)->getBytes();
            }

            if ($value instanceof Uuid) {
                return new Binary($value->getBytes(), Binary::TYPE_UUID);
            }

            throw new InvalidArgumentException(
                sprintf(
                    'Could not convert database value "%s" from "%s" to %s',
                    $value,
                    get_debug_type($value),
                    Binary::class
                )
            );
        }
    }

Register the type in your bootstrap code::

.. code-block:: php

    Type::addType(Ramsey\Uuid\Uuid::class, My\Project\Types\UuidType::class);

Usage Example::

.. code-block:: php

    #[Field(type: Ramsey\Uuid\Uuid::class)]
    public Ramsey\Uuid\Uuid $id;

By using the |FQCN| of the value object class as the type name, the type is
automatically used when encountering a property of that class. This means you
can omit the ``type`` option when defining the field mapping::

.. code-block:: php

    #[Field]

>     public Ramsey\Uuid\Uuid $id;

Author: This is a new feature that can compete with embedded documents: adding custom logic to convert a value object into BSON.

Author: Maybe the example should be with a custom class instead, since UUID are new supported using `symfony/uid` https://github.com/doctrine/mongodb-odm/pull/2826

        <field field-name="field" type="date_with_timezone" />

> Custom Type Example: Mapping a UUID Class

Reviewer 3: With UUID merged, WDYT of using [moneyphp](https://github.com/moneyphp/money) as an example?

Author: I updated the tutorial.

-----------------------------------------

You can create a custom mapping type for your own value objects or classes. For
example, to map a UUID value object using the `ramsey/uuid library`_, you can
implement a type that converts between your class and the BSON Binary UUID format.

This approach works for any custom class by adapting the conversion logic to your needs.

Example Implementation (using ``Ramsey\Uuid\Uuid``)::

.. code-block:: php

    <?php

    namespace App\MongoDB\Types;

    use Doctrine\ODM\MongoDB\Types\ClosureToPHP;
    use Doctrine\ODM\MongoDB\Types\Type;
    use InvalidArgumentException;
    use MongoDB\BSON\Binary as BsonBinary;
    use Ramsey\Uuid\Uuid as RamseyUuid;

    final class RamseyUuidType extends Type
    {
        // This trait provides default closureToPHP used during data hydration
        use ClosureToPHP;

        public function convertToPHPValue(mixed $value): ?RamseyUuid
        {
            if (null === $value) {
                return null;
            }

            if ($value instanceof BsonBinary && $value->getType() === BsonBinary::TYPE_UUID) {
                return RamseyUuid::fromBytes($value->getData());
            }

>             throw new \InvalidArgumentException(\sprintf('Could not convert database value from "%s" to %s',get_debug_type($value),RamseyUuid::class));

Reviewer 2: Should you add spaces after the argument-delimiting commas? You did that below.

Author: Good catch!
```suggestion
            throw new \InvalidArgumentException(\sprintf('Could not convert database value from "%s" to %s', get_debug_type($value), RamseyUuid::class));
```

        }

        public function convertToDatabaseValue(mixed $value): ?BsonBinary
        {
            if (null === $value) {
                return null;
            }

            if ($value instanceof RamseyUuid) {
                return new BsonBinary($value->getBytes(), BsonBinary::TYPE_UUID);
            }

            throw new \InvalidArgumentException(\sprintf('Could not convert database value from "%s" to %s', get_debug_type($value), Binary::class));
        }
    }

Register the type in your bootstrap code::

.. code-block:: php

    Type::addType(Ramsey\Uuid\Uuid::class, App\MongoDB\Types\RamseyUuidType::class);

Usage Example::

.. code-block:: php

    #[Field(type: \Ramsey\Uuid\Uuid::class)]
    public ?\Ramsey\Uuid\Uuid $id;

By using the |FQCN| of the value object class as the type name, the type is
automatically used when encountering a property of that class. This means you
can omit the ``type`` option when defining the field mapping::

.. code-block:: php

    #[Field]
    public ?\Ramsey\Uuid\Uuid $id;

.. note::

>     This implementation of ``RamseyUuidType`` is volontary simple and does not

Reviewer 2: I assume this is a typo for "voluntary" but I don't think that makes sense. Maybe say "intentionally simple"?

Author: ```suggestion
    This implementation of ``RamseyUuidType`` is intentionally simple and does not
```

                return new Money($value['amount'], new Currency($value['currency']));
            }

            throw new InvalidArgumentException(sprintf('Could not convert database value from "%s" to %s', get_debug_type($value), Money::class));
        }

        public function convertToDatabaseValue(mixed $value): ?array
        {
            if (null === $value) {
                return null;
            }

            if ($value instanceof Money) {
                return [
                    'amount' => $value->getAmount(),
                    'currency' => $value->getCurrency()->getCode(),
                ];
            }

            throw new InvalidArgumentException(sprintf('Could not convert database value from "%s" to array', get_debug_type($value)));
        }
    }

Register the type in your bootstrap code::

.. code-block:: php

    Type::addType(Money::class, App\MongoDB\Types\MoneyType::class);

Usage Example::

.. code-block:: php

    #[Field(type: \Money\Money::class)]
    public ?\Money\Money $price;

By using the |FQCN| of the value object class as the type name, the type is
automatically used when encountering a property of that class. This means you

> can omit the ``type`` option when defining the field mapping::

Reviewer 3: I would remove the first example and explain that the mapping logic automatically detects the property type and uses it as field type.

.. code-block:: php

    #[Field]
    public ?\Money\Money $price;

.. note::

>     This implementation of ``MoneyType`` is volontairely simple and does not

Reviewer 3: ```suggestion
    This implementation of ``MoneyType`` is kept simple for illustration purposes and does not
```
